Support Existing Azure AppInsights, OpenAI, and CosmosDB

eerhardt opened the pull request:
## Description

These resources don't support existing resources yet. Adding support to finish all the Azure resources for existing.

Note that for CosmosDB and Redis, we cannot explicitly set DisableLocalAuth on an existing resource because bicep doesn't support it.

Fix #7665

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
- Did you add public API?
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [x] No
- Does the change require an update in our Aspire docs?
  - [x] No

eerhardt committed:
Support Existing Azure AppInsights, OpenAI, and CosmosDB

These resources don't support existing resources yet. Adding support to finish all the Azure resources for existing.

Note that for CosmosDB, we only explicitly set DisableLocalAuth if WithAccessKeyAuthentication is called. We don't want to disable local auth on an existing resource if it is already enabled as that might break other uses of the resource.

Fix #7665

eerhardt commented on src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:
```diff
@@ -428,6 +434,28 @@ private static void ConfigureCosmosDBInfrastructure(AzureResourceInfrastructure
+{
+    public ExistingCosmosDBAccount(string bicepIdentifier, bool disableLocalAuth, string? resourceVersion = null) : base(bicepIdentifier, resourceVersion)
+    {
+        // only explicitly set DisableLocalAuth if WithAccessKeyAuthentication was called.
+        // We don't want to disable local auth on an existing resource if it is already enabled
+        // as that might break other uses of the resource.
+        if (disableLocalAuth is false)
+        {
+            DisableLocalAuth = false;
```
@captainsafia - I see for Redis that we always set it, is that intentional?

https://github.com/dotnet/aspire/blob/acdf5012407c2ab0259f4b6f075acb6c668949f3/src/Aspire.Hosting.Azure.Redis/AzureRedisExtensions.cs#L287-L294

https://github.com/dotnet/aspire/blob/acdf5012407c2ab0259f4b6f075acb6c668949f3/src/Aspire.Hosting.Azure.Redis/AzureRedisExtensions.cs#L205-L208

captainsafia replied:
Yes, this was intentional although you make a good point in your comment here and we probably want to constrain the Redis resource in the same way.

eerhardt replied:
It turns out this doesn't work end-to-end. If you set this property on an existing resource you get:

> Error BCP173: The property "properties" cannot be used in an existing resource declaration. [https://aka.ms/bicep/core-diagnostics#BCP173]

See https://github.com/Azure/azure-sdk-for-net/issues/48352#issuecomment-2669386450

eerhardt replied:
I've updated CosmosDB and Redis to no longer do this. I will make a follow up for Postgres since it has the same problem.

captainsafia commented on tests/Aspire.Hosting.Azure.Tests/ExistingAzureResourceTests.cs:
```diff
@@ -1254,6 +1275,314 @@ param keyVaultName string
             }
             """;
 
+        output.WriteLine(BicepText);
```
Is this just here to help with test output?

eerhardt replied:
It's there to help when the bicep changes and you need to know what changed. You can copy the actual bicep from the test log and paste it into the test code. That will show you all the differences. This is better/easier than getting the diffs one at a time from the test failure:

![image](https://github.com/user-attachments/assets/78370b10-aa47-440f-ad67-1ec673478fde)

eerhardt commented on tests/Aspire.Hosting.Azure.Tests/ExistingAzureResourceTests.cs:
```diff
@@ -134,6 +138,8 @@ param principalId string
             output serviceBusEndpoint string = messaging.properties.serviceBusEndpoint
             """;
+
+        output.WriteLine(BicepText);
```
It isn't redundant. It is helpful to diagnose changes in behavior.

eerhardt committed:
Remove setting disable local auth on existing resources - CosmosDB and Redis

Existing Azure resources can't be directly updated in bicep. When trying to deploy these changes it raises an error "Error BCP173: The property "properties" cannot be used in an existing resource declaration. [https://aka.ms/bicep/core-diagnostics#BCP173]"

captainsafia commented on src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:
```diff
@@ -324,26 +324,34 @@ public static IResourceBuilder<AzureCosmosDBResource> WithAccessKeyAuthenticatio
-            Tags = { { "aspire-resource-name", infrastructure.AspireResource.Name } }
-        };
-        infrastructure.Add(cosmosAccount);
+                DisableLocalAuth = disableLocalAuth,
```
+1 to this. Setting this directly on the resource instead of after the fact reads better as well.

eerhardt committed:
Fixing AppInsights test. The resources were reordered because they are added to the infrastructure in the opposite order now.

eerhardt commented:
/backport to release/9.1
